src/main/java/org/opensearch/flowframework/template/ProcessNode.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.flowframework.template;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.flowframework.workflow.WorkflowData;
+import org.opensearch.flowframework.workflow.WorkflowStep;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Representation of a process node in a workflow graph.  Tracks predecessor nodes which must be completed before it can start execution.
+ */
+public class ProcessNode {
+
+    private static final Logger logger = LogManager.getLogger(ProcessNode.class);
+
+    private final String id;
+    private final WorkflowStep workflowStep;
+    private final WorkflowData input;
+    private CompletableFuture<WorkflowData> future = null;
+
+    // will be populated during graph parsing
+    private Set<ProcessNode> predecessors = Collections.emptySet();
+
+    /**
+     * Create this node linked to its executing process.
+     *
+     * @param id A string identifying the workflow step
+     * @param workflowStep A java class implementing {@link WorkflowStep} to be executed when it's this node's turn.
+     */
+    ProcessNode(String id, WorkflowStep workflowStep) {
+        this(id, workflowStep, WorkflowData.EMPTY);
+    }
+
+    /**
+     * Create this node linked to its executing process.
+     *
+     * @param id A string identifying the workflow step
+     * @param workflowStep A java class implementing {@link WorkflowStep} to be executed when it's this node's turn.
+     * @param input Input required by the node
+     */
+    public ProcessNode(String id, WorkflowStep workflowStep, WorkflowData input) {
+        this.id = id;
+        this.workflowStep = workflowStep;
+        this.input = input;
+    }
+
+    /**
+     * Returns the node's id.
```

Comment:
is this node id information coming from use-case template or it will be assigned at backend? I don't remember seeing this node id on the use-case template, you might want to confirm with front end people if you need this information from use-case template

Reply:
There is currently an `id` field with a somewhat descriptive name in the proposed use case template, which is what I based the JSON for this on. 

It needs to be unique, which we haven't really explicitly stated, but it's the same string that is used to define the source and destination for the edges, which somewhat forces that requirement.